code/datums/job_selection/job_selector.dm

```diff
@@ -0,0 +1,320 @@
+RESTRICT_TYPE(/datum/job_selector)
+
+/datum/job_selector
+	var/list/candidates = list()
+	var/list/assigned_candidates = list()
+	var/probability_of_antag_role_restriction = 100
+
+/datum/job_selector/proc/Debug(text)
+#ifdef GAME_TESTS
+	log_chat_debug(text)
+#else
+	SSjobs.Debug(text)
+#endif
+
+/datum/job_selector/proc/add_candidate(mob/new_player/player = null)
+	var/datum/job_candidate/candidate = new()
+	if(player)
+		candidate.load_from_player(player)
+
+	candidates[candidate] = player
+
+/datum/job_selector/proc/assign_role(datum/job_candidate/candidate, datum/job/job, latejoin = FALSE)
+	if(!job)
+		return FALSE
+
+	var/eligible = candidate.get_job_eligibility(job)
+	var/available = latejoin ? job.is_position_available() : job.is_spawn_position_available()
+
+	if(eligible && available)
+		candidate.assign_title(job.title)
+		assigned_candidates[candidate] = candidates[candidate]
+		candidates -= candidate
+		job.current_positions++
+		SSblackbox.record_feedback("nested tally", "manifest", 1, list(job.title, (latejoin ? "latejoin" : "roundstart")))
+		return TRUE
+
+	return FALSE
+
+/// Convenience proc for handling a single latejoin player
+/datum/job_selector/proc/latejoin_assign(mob/new_player/player, datum/job/job)
+	var/datum/job_candidate/candidate = new()
+	candidate.load_from_player(player)
+	if(assign_role(candidate, job, latejoin = TRUE))
+		candidate.apply_to_player(player)
+	else
+		to_chat(player, "<span class='warning'>You are unable to join the round as [job.title]. Please try another job.</span>")
+
+/datum/job_selector/proc/apply_roles_to_players()
+	for(var/datum/job_candidate/candidate in assigned_candidates)
+		var/mob/new_player/player = assigned_candidates[candidate]
+		if(player)
+			candidate.apply_to_player(player)
+
+/datum/job_selector/proc/return_to_lobby(datum/job_candidate/candidate)
+	candidate.return_to_lobby()
+	assigned_candidates[candidate] = candidates[candidate]
+	candidates -= candidate
+
+/datum/job_selector/proc/find_job_candidates(datum/job/job, level, flag)
+	var/list/job_candidates = list()
+	for(var/datum/job_candidate/candidate in candidates)
```

Comment:
same here, and for a few of these other loops -- you can probably as anything them for some micro-optimizations

Reply:
done.